src/daemon/rpmostreed-os-experimental.cxx

```diff
@@ -207,30 +208,53 @@ osexperimental_handle_download_packages (RPMOSTreeOSExperimental *interface,
     goto out;
 
   sack = dnf_context_get_sack (rpmostree_context_get_dnf (ctx));
+  RpmOstreeOverrideSourceKind source_kind;
+  source_name = rpmostree_parse_override_source (source, &source_kind, &local_error);
 
-  for (const char* const* it = queries; it && *it; it++)
-    {
-      auto query = static_cast<const char*> (*it);
-      g_autoptr(GPtrArray) pkglist = rpmostree_get_matching_packages (sack, query);
+  if (local_error)
+    goto out;
 
-      if (!pkglist || pkglist->len == 0)
+  if (source_kind == RPM_OSTREE_OVERRIDE_SOURCE_KIND_REPO)
+    {
+      for (const char* const* it = queries; it && *it; it++)
         {
-          glnx_throw (&local_error, "No matching packages for \"%s\"", query);
-          goto out;
+          auto pkg_name = static_cast<const char*> (*it);
+          g_autoptr(GPtrArray) pkglist = NULL;
+          HyNevra nevra = NULL;
+          g_auto(HySubject) subject = hy_subject_create(pkg_name);
+          hy_autoquery HyQuery query =
+            hy_subject_get_best_solution(subject, sack, NULL, &nevra, FALSE, TRUE, TRUE, TRUE, FALSE);
+          hy_query_filter (query, HY_PKG_REPONAME, HY_EQ, source_name);
+          pkglist = hy_query_run (query);
+          if (!pkglist || pkglist->len == 0)
+            {
+              glnx_throw (&local_error, "No matches for \"%s\" in repo '%s'", pkg_name, source_name);
+              goto out;
+            }
+          g_ptr_array_add (dnf_pkgs, g_object_ref (pkglist->pdata[0]));
         }
-      g_ptr_array_add (dnf_pkgs, g_object_ref (pkglist->pdata[0]));
+    }
+  /* Future source kinds go here */
+  else
+    {
+       glnx_throw (&local_error, "Unsupported source type: %s", source);
+       goto out;
     }
 
+  if (!dnf_pkgs || dnf_pkgs->len == 0)
+    goto out;
```

Comment:
The call to this function is preceded by this line 122 at `src/app/rpmostree-override-builtins.cxx`:
`      if (queries->len > 0)` 
Does this mean we can just drop the check at `src/daemon/rpmostreed-os-experimental.cxx` because there is no way the string array could be empty?

Reply:
We should still check it on the daemon side, because the D-Bus interface can be used by more than just the CLI.